src/doc/src/reference/manifest.md

```diff
@@ -312,7 +312,7 @@ overflow-checks = true # use overflow checks for integer arithmetic.
 opt-level = 3
 debug = false
 rpath = false
-lto = 'thin'
+lto = false
```

Comment:
@alexcrichton would now better, but shouldn't we use thin lto at least for release builds? Like, we have `codegen-units`...

Reply:
I thought what currently happens (in nightly) is

release build: enable thin lto by default
debug build: don't enable any link time optimizations

[profile.release] in cargo toml:

lto = "true" enables monolithic lto
no lto option: enables thin lto for release build (since its default)
lto="thin":  also enables thin lto as well.

Is that wrong?

Reply:
Yep, this is what I think as well. So, for `profile.release` `lto=thin` is default, and not `lto=false`. So `lto = 'thin'` is **correct** in this particular instance? I think in this place we list default settings for profiles.

Reply:
The distinctions here are pretty subtle. ThinLTO basically is a scalable/incremental way of saying "pretend these N translation units are one translation unit". The default release mode of the compiler is to intend to interpret each crate as if it were one translation unit and optimize accordingly (modulo actual implementation details). To that end we have indeed turned on multiple codegen units by default on nightly for the parallelization wins, and to retain the same level of performance of "pretend the crate is one translation unit" we've enabled ThinLTO amongst *only that crate's codegen units*.

The `lto` option in the manifest is currently interpreted as "if not `false`, pretend the entire crate graph is one translation unit". For `lto = true` that literally happens (everything goes in one translation unit). For `lto = 'thin'` it runs ThinLTO passes over all objects in the crate graph. This is *much* faster to compile and has the possibility for incrementally working as opposed to `lto = 'fat'`.

So the tl;dr; here is that it's whether you want to think of the crate as one translation unit or the entire crate graph as one translation unit. The compiler's default release mode will always consider at least a crate as the translation unit (but uses ThinLTO today as an implementation detail to compile faster), and the `lto` option in the manifest is instructing the compiler "no please expand your view to the entire crate graph".

The name "ThinLTO" is probably not the best for what rustc is doing, and I'm pretty sure we're not using the ThinLTO code generation as it was intended to be used. In that sense what we're doing in rustc is somewhat nonstandard.

Reply:
Interesting! Thanks a lot for this explanation!

I understand why we don't use the fastest `lto='fat'` for release profile: it's prohibitively slow to compile. But I wonder if it makes sense to *change* the default for release profile to `lto='thin'`? Or is it still slow and not 100% reliable?

Reply:
Whole crate graph ThinLTO is still much slower than only running ThinLTO on the local crate, so turning that on would probably cause too much of a compile time regression for too little benefit.

Reply:
Is ````lto=true```` equivalent to ````lto="fat"```` ?

Reply:
Indeed!

`lto=true` means `-C lto` and `lto = 'fat'` means `-C lto=fat`, and rustc interprets those the same way